# Get minor allele frequency estimations

Now that we have a database of global snps we're going to get minor alleles frequencies for each population

#!/bin/bash

#SBATCH --job-name popmafs_angsd
#SBATCH --output %A_popmafs_angsd.out
#SBATCH --error %A_popmafs_angsd.err
#SBATCH --mail-type=ALL
#SBATCH --mail-user=mabrober@ucsc.edu
#SBATCH --time=15-00:00:00
#SBATCH --ntasks=24
#SBATCH --mem=120GB

## This script is used to get minor allele frequency estimation from angsd for each population / group

module load angsd

BASEDIR=/hb/groups/bernardi_lab/may/DTR/population-analysis/ #  Path to the base directory where adapter clipped fastq file are stored in a subdirectory titled "adapter_clipped" and into which output files will be written to separate subdirectories. An example for the Greenland cod data is: /workdir/cod/greenland-cod/
SAMPLETABLE=/hb/groups/bernardi_lab/may/DTR/population-analysis/sample_lists/clean_sample_table_merged.tsv # Path to a sample table where the 1st column is the prefix of the raw fastq files. The 4th column is the sample ID, the 2nd column is the lane number, and the 3rd column is sequence ID. The combination of these three columns have to be unique. The 6th column should be data type, which is either pe or se. An example of such a sample table is: /workdir/cod/greenland-cod/sample_lists/sample_table.tsv
POPCOLUMN=5 # The column index of the variable that you want to group by in the sample table above. In the Greenland project, it's the fifth column,  and thus 5
#BAMLISTPREFIX=$4 # Prefix of the bam lists. An example from the Greenland cod project is bam_list_realigned_
REFERENCE=/hb/groups/bernardi_lab/may/DTR/DTRgenome/genome_annot/genome/kuro_filt_s500.fasta # Path to reference genome
SNPLIST=/hb/groups/bernardi_lab/may/DTR/population-analysis/angsd/global_snp_list_DTR_dedup_bams_mindp100_maxdp590_minind13_minq20.txt # Path to the SNP list
MINDP=6 # Minimum depth filter #same as MININD
MAXDP=14 # Maximum depth filter #MINDIND x estimated coverage (2.3)
MININD=6 # Minimum individual filter # 2/3 the number of your smallest population size (9)
MINQ=20 # Minimum quality filter
MINMAPQ=20 # Minimum mapping quality (alignment score) filter, default value is 20
#ANGSD=${12:-/workdir/programs/angsd0.931/angsd/angsd} # Path to ANGSD, default value is /workdir/programs/angsd0.931/angsd/angsd
THREADS=24 # Number of parallel threads to use, default value is 8.
EXTRA_ARG='-remove_bads 1 -only_proper_pairs 1 -C 50' # Extra arguments when running ANGSD, default value is '-remove_bads 1 -only_proper_pairs 1 -C 50'

OUTBASE=`echo $SNPLIST | sed 's/\..*//' | sed -e 's#.*\/\(\)#\1#'`
CHRLIST=`echo $SNPLIST | sed 's/\..*//'`.chrs
OUTDIR=$BASEDIR'angsd/popminind'$MININD'/'
if [ ! -d "$OUTDIR" ]; then
				mkdir $OUTDIR
fi

for POP in `tail -n +2 $SAMPLETABLE | cut -f $POPCOLUMN | sort | uniq`; do
				echo $POP
				angsd \
				-b $BASEDIR'sample_lists/bam_list_per_pop/'$POP'.txt' \
				-anc $REFERENCE \
				-ref $REFERENCE \
				-out $OUTDIR$POP'_'$OUTBASE'_popminind'$MININD \
				-dosaf 1 -GL 1 -doGlf 2 -doMaf 1 -doMajorMinor 3 -doCounts 1 -doDepth 1 -dumpCounts 1 \
				-P $THREADS \
				-setMinDepth $MINDP -setMaxDepth $MAXDP -minInd $MININD -minQ $MINQ -minMapQ $MINMAPQ \
				-sites $SNPLIST -rf $CHRLIST \
				$EXTRA_ARG \
				>& $BASEDIR'angsd/'$POP'_'$OUTBASE'_popminind'$MININD'_maf.log'
done
