# Get minor allele frequency estimations

Now that we have a database of global snps we're going to get minor alleles frequencies for each population. 

## Make population bam lists
First, we need to make all the lists of bams for each population, ie a list of paths to all bam files for population Amami Oshima called AMA.txt\

`3.1-make_pop_lists.sh` Ran this on the head node ~1sec

	LISTDIR=/hb/groups/bernardi_lab/may/DTR/population-analysis/sample_lists/
	LIST=DTR_dedup_bams.tsv
	SAMPLETABLE=/hb/groups/bernardi_lab/may/DTR/population-analysis/sample_lists/clean_sample_table_merged.tsv
	POPCOLUMN=5 # The column index of the variable that you want to group by in the sample table above.

	for POP in `tail -n +2 $SAMPLETABLE | cut -f $POPCOLUMN | sort | uniq`; do
		grep "$POP" $LISTDIR$LIST > $LISTDIR/bam_list_per_pop/$POP.txt
	done

## Get mafs

This script will output 9 files into a diectory that it creats called popminind$MININD

`3.2-get_maf_per_pop_angsd.mpi`

	#!/bin/bash

	#SBATCH --job-name popmafs_angsd
	#SBATCH --output %A_popmafs_angsd.out
	#SBATCH --error %A_popmafs_angsd.err
	#SBATCH --mail-type=ALL
	#SBATCH --mail-user=mabrober@ucsc.edu
	#SBATCH --partition=128x24
	#SBATCH --nodes=1
	#SBATCH --time=30-00:00:00
	#SBATCH --mem=120GB
	#SBATCH --ntasks-per-node=20

	## This script is used to get minor allele frequency estimation from angsd for each population / group

	module load angsd

	BASEDIR=/hb/groups/bernardi_lab/may/DTR/population-analysis/
	SAMPLETABLE=/hb/groups/bernardi_lab/may/DTR/population-analysis/sample_lists/clean_sample_table_merged.tsv # Path to the sample table we've been using
	POPCOLUMN=5 # The column index of the variable that you want to group by in the sample table above. 
	REFERENCE=/hb/groups/bernardi_lab/may/DTR/DTRgenome/genome_annot/genome/kuro_filt_s500.fasta # Path to reference genome
	SNPLIST=/hb/groups/bernardi_lab/may/DTR/population-analysis/angsd/global_snp_list_DTR_dedup_bams_mindp100_maxdp590_minind13_minq20.txt # Path to the SNP list
	MINDP=6 # Minimum depth filter #same as MININD
	MAXDP=14 # Maximum depth filter #MINDIND x estimated coverage (2.3)
	MININD=6 # Minimum individual filter # 2/3 the number of your smallest population size (9)
	MINQ=20 # Minimum quality filter
	MINMAPQ=20 # Minimum mapping quality (alignment score) filter, default value is 20
	THREADS=24 # Number of parallel threads to use, default value is 8.

	OUTBASE=`echo $SNPLIST | sed 's/\..*//' | sed -e 's#.*\/\(\)#\1#'`
	CHRLIST=`echo $SNPLIST | sed 's/\..*//'`.chrs
	OUTDIR=$BASEDIR'angsd/popminind'$MININD'/'
	if [ ! -d "$OUTDIR" ]; then
					mkdir $OUTDIR
	fi

	for POP in `tail -n +2 $SAMPLETABLE | cut -f $POPCOLUMN | sort | uniq`; do
		echo $POP
		angsd \
		-b $BASEDIR'sample_lists/bam_list_per_pop/'$POP'.txt' \
		-anc $REFERENCE \
		-ref $REFERENCE \
		-out $OUTDIR$POP'_'$OUTBASE'_popminind'$MININD \
		-dosaf 1 -GL 1 -doGlf 2 -doMaf 1 -doMajorMinor 3 -doCounts 1 -doDepth 1 -dumpCounts 1 \
		-P $THREADS \
		-setMinDepth $MINDP -setMaxDepth $MAXDP -minInd $MININD -minQ $MINQ -minMapQ $MINMAPQ \
		-sites $SNPLIST -rf $CHRLIST \
		$EXTRA_ARG \
		>& $BASEDIR'angsd/'$POP'_'$OUTBASE'_popminind'$MININD'_maf.log'
	done
